### ISSUE TITLE
server: integrate the TLS auto-negotiation inside the crdb code

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#60632]
Fixes #32448. 
Epic: CRDB-6663
Jira issue: CRDB-3167

Meta-issue to track the implementation for #51991

A couple of new components:

- existing CM manager remains unchanged, not involved in the discussion here
- [x] new cert generation primitives in `security` package #60705 
- [x] initial cert handshake HTTP server
- [x] initial cert handshake HTTP client
- [x] cert creation loop with check of which CAs/certs are already provided by the operator #60705 
- [x] ~(maybe in later phase) extend the cert rotation code with automatic handshake of the new batch of certs~
- [x]  join token generation and delivering a cert bundle for new nodes being added  #63492
- [x] security: Add version number to base64 form of join tokens #64885
- [ ] end-to-end TCL unit test for the node join functionality
- [ ] Have connect commands use a configurable timeout #64878
- [ ] mechanism to trigger the rotation automatically
- [ ] host cert  + init token check during `start` to decide whether to engage `connect` or start as usual  #63850
- [ ] cli/connect: make join warn/error the user if the target directory already contains certs #64944

Follow-up work:

- [ ] ping the folk on #38539 to tell them to use the new code in roachprod
- [ ] server: clarify / improve the TLS join behavior for already-running, not-yet-initialized clusters #64934

Clean-up work:

- [ ] security: Reduce code duplication between auto_tls_init and the rest of the security package #64883
- [ ] security: Have TLS auto-join/init code use CertificateManager #64884

CLI commands:

- [x] `cockroach connect` - new, only responsible for TLS handshake and writing the certs to disk
  This will leverage the first 3 components identified above: cert gen primitives + HTTP client/server for handshake.
- [ ] `cockroach start` - when provided an init token, must check if the host cert is known already and if not start the TLS handshake before the remainder of the start code  #63850
- [ ] `cockroach start-single-node` - new flag `--self-secure-init` that auto-generates an init token and proceeds as per the `start` logic
- [ ] `cockroach demo` - will be modified to leverage the self secure init code added to start-single-node 
- [ ]  Add `--init-token-file` flag to protect the init handshake shared secret #61231

Bugs:

- [ ] Nodes should populate handshake messages with advertise address instead of listen #61238 
- [ ] `node.key` permissions are incorrect with a default umask
- [ ]  cockroach connect join gets confused by CA key file format #64942 #64943
- [ ] flaky test #61538

Technical question where the answer is needed as prereq to a number of points above:

- how are the CN and OU fields populated
- how is the SAN field populated
  - current assumption for prototype/MVP: the addresses provided on `--join` go into the SAN? or maybe `--listen-addr`? (Unsure, this is under-specified)
  - need a practical test with a multi-server experiment, to understand the design constraint
  - there may be some flags / extra logic needed to pick up reasonable + valid addresses to populate SAN

Possible action item: perform that experiment

**Comment by knz** [comment]
cc @aaron-crl @itsbilal - can you check this list and add your own items ahead of our meeting later today

Some left over items to integrate in the to-do list:

- `cockroach connect` should honor the CN override recognized by `cockroach cert` (via env var)
- drop the `--certs-dir` requirement for `start` commands
- with a default `uname`, I get the following error upon `start`:
  ```
  failed to start server: problem using security settings: key file logs/sn/node.key has permissions -rw-rw-rw-, exceeds -rwx------
  ```
- see issue #61238 
- see issue #61231

**Comment by aaron-crl** [comment]
We need to check status of #61230 once #60854 lands. I _think_ #61233 may have addressed this with updated defaults.

**Comment by dt** [comment]
As discussed elsewhere, I think it might be worth considering switching to an off-the-shelf encoder like proto to handle join token encoding/decoding before we get locked into compatibility with a shipped scheme: https://github.com/cockroachdb/cockroach/issues/65344

**Comment by knz** [comment]
Tracking the list of items here as a jira epic: CRDB-6663.
